### PR TITLE
pyflyte register

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Call it like so:
```
~/go/src/github.com/flyteorg/flytesnacks/cookbook/core [flyte-sandbox] (master) $ pyflyte register -v abc12 --service-account jfkldsje --raw-data-prefix s3://bob/var flyte_basics/basic_workflow.py control_flow/
```

always fast register for now.  If no version is given one is computed.

will change warnings to debug after unit tests

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

### Usage
Can use a combination of both folders and files.

### Others
* Removing a bunch of unused functions in `module_loader`
* Move the version hashing logic into a separate function.
* Add the same tricks to the `fast_package` function as we did for a single script so that the tar.gz file hashes consistently.
* When serializing, pass Options along when getting the default launch plan.

### Minor fixes
* In translator, first check the raw output data in Options, if provided.
* Fix some type hints.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2541

## Follow-up issue
https://github.com/flyteorg/flyte/issues/2540

### CODE REVIEW

File: flytekit/clis/sdk_in_container/register.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    image_config: ImageConfig,
+    output: str,
+    destination_dir: str,
+    service_account: str,
+    raw_data_prefix: str,
+    version: typing.Optional[str],
+    package_or_module: typing.Tuple[str],
+):
+    """
+    This command produces a Flyte backend registrable package of all entities in Flyte.
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/register.py:102]: this is wrong right? It seems like serialize

File: flytekit/clis/sdk_in_container/register.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        serialization_settings, detected_root, list(package_or_module), options
+    )
+    cli_logger.warning(f"Found and serialized {len(registerable_entities)} entities")
+
+    if not version:
+        version = remote._version_from_hash(md5_bytes, serialization_settings, service_account, raw_data_prefix)  # noqa
+        cli_logger.warning(f"Computed version is {version}")
+
+    # Register using repo code
+    repo_register(registerable_entities, project, domain, version, remote.client)
```
Comment (kumare3) [inline flytekit/clis/sdk_in_container/register.py:161]: is this the same as package? I hope so right

Comment (wild-endeavor) [reply flytekit/clis/sdk_in_container/register.py:161]: yes i'm importing it from there

File: flytekit/clis/sdk_in_container/register.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    Note that this command runs "fast" register by default. Future work to come to add a non-fast version.
+    This means that a zip is created from the detected root of the packages given, and uploaded. Just like with
+    pyflyte run, tasks registered from this command will download and unzip that code package before running.
+    """
+
+    pkgs = ctx.obj[constants.CTX_PACKAGES]
+    if not pkgs:
+        cli_logger.debug("No pkgs")
+    if pkgs:
+        raise ValueError("to do, please implement")
```
Comment (eapolinario) [inline flytekit/clis/sdk_in_container/register.py:116]: how would that look like?

**Review by eapolinario — COMMENTED** [review]
Left a bunch of comments, nothing major though. Thanks for getting this out, this is looking amazing!

File: flytekit/remote/remote.py [hunk 4]
```diff
@@ -507,6 +508,66 @@ def register_workflow(
...
+        :param project: Project to upload under, if not supplied will use the remote's default
+        :param domain: Domain to upload under, if not specified will use the remote's default
+        :return: The uploaded location.
+        """
+        if not to_upload.is_file():
+            raise ValueError(f"{to_upload} is not a single file, upload arg must be a single file.")
+        md5_bytes, str_digest = hash_file(to_upload)
+        remote_logger.debug(f"Text hash of file to upload is {str_digest}")
+
+        c = self.client
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/remote/remote.py:527]: oops

File: flytekit/tools/ignore.py [hunk 5]
```diff
@@ -105,7 +105,7 @@ class IgnoreGroup(Ignore):
     """Groups multiple Ignores and checks a path against them. A file is ignored if any
     Ignore considers it ignored."""
 
-    def __init__(self, root: str, ignores: List[Ignore]):
+    def __init__(self, root: str, ignores: List[Type[Ignore]]):
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/tools/ignore.py:108]: oh it was wrong.  it's called like `ignore = IgnoreGroup(source, [GitIgnore, DockerIgnore, StandardIgnore])`  the classes are passed, not instances of the class.

File: flytekit/tools/module_loader.py [hunk 6]
```diff
@@ -36,37 +23,14 @@ def just_load_modules(pkgs: List[str]):
...
+        # If it doesn't have a __path__ field, that means it's not a package, just a module
+        if not hasattr(package, "__path__"):
+            continue
 
-    if additional_path is not None:
-        with add_sys_path(additional_path):
-            return iterate()
-    else:
-        return iterate()
+        # Note that walk_packages takes an onerror arg and swallows import errors silently otherwise
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/tools/module_loader.py:31]: nah... i think it's fine.  we've never had an issue.  i think we need to take a look at loading in general one day soon too, maybe we can get to it then.

File: flytekit/clis/sdk_in_container/register.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/clis/sdk_in_container/register.py:116]: probably use in lieu of the args provided?  maybe... let me change the text.  maybe we just leave it unimplemented.

File: tests/flytekit/unit/tools/test_repo.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        with open(os.path.join(middle_level, "__init__.py"), "w"):
+            pass
+        with open(os.path.join(middle_level, "a.py"), "w"):
+            pass
+        with open(os.path.join(bottom_level, "__init__.py"), "w"):
+            pass
+        with open(os.path.join(bottom_level, "a.py"), "w"):
+            pass
+        with open(os.path.join(middle_level_2, "__init__.py"), "w"):
+            pass
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply tests/flytekit/unit/tools/test_repo.py:42]: done

File: flytekit/clis/sdk_in_container/register.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        domain=domain,
+        image_config=image_config,
+        fast_serialization_settings=FastSerializationSettings(
+            enabled=True,
+            destination_dir=destination_dir,
+            distribution_location=native_url,
+        ),
+    )
+
+    options = Options.default_from(k8s_service_account=service_account, raw_data_prefix=raw_data_prefix)
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/clis/sdk_in_container/register.py:151]: it should but will leave this for another PR... the difference is that before we pass None whereas if we don't check we pass Options.default_from.  given the amount of issues we've had around this i'd rather do this in a separate PR.

File: flytekit/tools/repo.py [hunk 9]
```diff
@@ -113,41 +121,100 @@ def register(
...
-    source: str = ".",
-    fast: bool = False,
 ):
-    if fast:
-        # TODO handle fast
-        raise AssertionError("Fast not handled yet!")
-    for entity, cp_entity in registrable_entities:
+    # The incoming registrable entities are already in base protobuf form, not model form, so we use the
+    # raw client's methods instead of the friendly client's methods by calling super
+    for admin_entity in registrable_entities:
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/tools/repo.py:127]: adding a check at the click level.  i think it looks better there.

File: tests/flytekit/unit/tools/test_repo.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        serialization_settings = flytekit.configuration.SerializationSettings(
+            project="project",
+            domain="domain",
+            version="version",
+            env=None,
+            image_config=ImageConfig.auto(img_name=DefaultImages.default_image()),
+        )
+
+        # Not a good test but at least try to load
+        load_packages_and_modules(serialization_settings, pathlib.Path(root), [bottom_level])
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply tests/flytekit/unit/tools/test_repo.py]: done

File: flytekit/clis/sdk_in_container/register.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    detected_root = find_common_root(package_or_module)
+    cli_logger.debug(f"Using {detected_root} as root folder for project")
+    zip_file = fast_package(detected_root, output)
+
+    # Upload zip file to Admin using FlyteRemote.
+    md5_bytes, native_url = remote._upload_file(pathlib.Path(zip_file))
+    cli_logger.debug(f"Uploaded zip {zip_file} to {native_url}")
+
+    # Create serialization settings
+    # Todo: Rely on default Python interpreter for now, this will break custom Spark containers
```
[the remark this answers is not in the document]
Comment (eapolinario) [reply flytekit/clis/sdk_in_container/register.py:146]: Actually, we set the python interpreter used by spark [here](https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-spark/flytekitplugins/spark/task.py#L99) and that is then used [in the spark plugin](https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/k8s/spark/spark.go#L141-L144) to configure the python version used.

**Review by eapolinario — APPROVED** [review]
Amazing!
